File: 2016/nim/day1/day1.nim
import sets
import strformat
import strutils

var
  moves = readAll(stdin).strip()
  dir = 'N'
  x = 0
  y = 0
  known_places = toSet([[0, 0]])
  current_place = [0, 0]
  found_place = false
  turn: char
  steps: int

for move in moves.split(", "):
  turn = move[0]
  steps = parseInt(move[1 .. move.high])

  if   (dir, turn) == ('N', 'L'):
    dir = 'W'
  elif (dir, turn) == ('N', 'R'):
    dir = 'E'
  elif (dir, turn) == ('W', 'L'):
    dir = 'S'
  elif (dir, turn) == ('W', 'R'):
    dir = 'N'
  elif (dir, turn) == ('S', 'L'):
    dir = 'E'
  elif (dir, turn) == ('S', 'R'):
    dir = 'W'
  elif (dir, turn) == ('E', 'L'):
    dir = 'N'
  elif (dir, turn) == ('E', 'R'):
    dir = 'S'

  for _ in 1 .. steps:
    if   dir == 'N':
      y += 1
    elif dir == 'W':
      x -= 1
    elif dir == 'S':
      y -= 1
    elif dir == 'E':
      x += 1

    current_place = [x, y]
    if not found_place and known_places.contains(current_place):
      echo "qb: ", abs(x) + abs(y)
      found_place = true
    else:
      known_places.incl(current_place)

echo "qa: ", abs(x) + abs(y)

File: 2016/julia/day1/day1.jl
function main(x, y, dir, moves)
    found_place = false
    known_places = Set([[x, y]])

    for move in moves
        turn = move[1]
        steps = parse(Int32, move[2:end])

        if     dir == 'N' && turn == 'L'
            dir = 'W'
        elseif dir == 'N' && turn == 'R'
            dir = 'E'
        elseif dir == 'W' && turn == 'L'
            dir = 'S'
        elseif dir == 'W' && turn == 'R'
            dir = 'N'
        elseif dir == 'S' && turn == 'L'
            dir = 'E'
        elseif dir == 'S' && turn == 'R'
            dir = 'W'
        elseif dir == 'E' && turn == 'L'
            dir = 'N'
        elseif dir == 'E' && turn == 'R'
            dir = 'S'
        end

        for _ in 1:steps
            if     (dir == 'N')
                y += 1
            elseif (dir == 'W')
                x -= 1
            elseif (dir == 'S')
                y -= 1
            elseif (dir == 'E')
                x += 1
            end

            current_place = [x, y]
            if !found_place && current_place in known_places
                found_place = true
                println("qb: ", abs(x) + abs(y))
            end

            known_places = union(known_places, Set([current_place]))
        end
    end

    println("qa: ", abs(x) + abs(y))
end

content = strip(read(stdin, String))
moves = split(content, ", ")
main(0, 0, 'N', moves)
